prebid-server/pbs-modules/rules-engine.md

```diff
@@ -0,0 +1,805 @@
+
+All of these functions operate only on the OpenRTB request at the 'Processed Auction Request' stage of module invocation.
+
+{: .table .table-bordered .table-striped }
+| Schema Function | Args | Description | Output Example |
+| :---- | :---- | :---- | :---- |
+| deviceCountry | null  | Returns OpenRTB device.geo.country. | "CAN" |
+| deviceCountryIn | array of strings representing country codes | If the parameter is an array of strings, then check to see if device.geo.country is part of the named set (case sensitive). If yes, then return "true", else return "false". The parameter must be an array of strings. No length limit. If the array is length zero, treat it as 'null' above. | "true" |
+| datacenters | null  This is an option for those who don't have geo-lookup available on every request. | Returns the value of the datacenter, which is host-company specific. Please work with your host company to get the list of values available in your environment. | "eu-west" |
+| datacentersIn | array of strings containing PBS host company-specific code for region.  This is an option for those who don't have geo-lookup available on every request. | Arguments must be an array of strings with no length limit. The function returns "true"/"false" if the actual datacenter is on the array (case sensitive). Please work with your host company to get the list values available in your environment. | "true" |
+| channel | null | Returns ext.prebid.channel – e.g. web, amp, app. Returns empty string if there is no channel. | "app" |
+| eidAvailable | null  | Returns "true" if user.eids array exists and is non-empty. Returns "false" if user.eids doesn't exist or is empty. | "false" |
+| eidIn | array of strings meant to be extended ID source values like "pubcid.org" | Returns "true"/"false" if the named eid.source is on the array (case sensitive). | "true" |
+| userFpdAvailable | null | Scans user.ext.data and user.data  Returns "true" if either user.ext.data or user.data exists and is non-empty. If neither exists with data, returns "false". Maybe be extended in the future to support specific fields. | "false" |
+| fpdAvailable | null | Returns "true" if any of the following objects exists and has data: user.ext.data, user.data, site.ext.data, site.content.data, app.ext.data, app.content.data. Otherwise returns "false". | "false" |
+| gppSidAvailable | null | Returns "true" if there are any values on regs.gpp\_sid greater than 0. Otherwise returns "false". | "true" |
+| gppSidIn | array of ints | Returns "true" if any of the regs.gpp\_sid values are on the supplied array. (i.e. intersection is true). Otherwise returns "false". | "true" |
```

Comment:
same comment as `deviceCountryIn` regarding array of strings with at least one element